### DISCUSSION TITLE
Jib.reqMktData throws   e = TypeError: in typeassert, expected Float64, got a value of type Nothing

### DISCUSSION THREAD

**Discussion by ghost** [lbilli/Jib.jl#17] [Q&A] [answered]
Hi there
remember I am a noob so please bear with me.I am trying not to bother you.

 I added a 10 sec sleep and made the method override to see how it works but I got an error I don't know how to diagnose. I am running all of this in the repl and text editor. 

ALSO I feel bad about raising these things as issues, do you have a julia discourse thread please?

```
using Jib
symbols = ["AAPL", "TSLA", "AMD", "INTC"]


ib = Jib.connect("192.168.1.219",1234, 21)

d, wrap = Jib.simple_wrap();

wrap.tickPrice = (tickerId, field, price, size, attrib) -> println("tickPrice2: $(symbols[tickerId]) $field $price $size $attrib")

Jib.start_reader(ib, wrap)

for (idx, s) in enumerate(symbols)
  contract = Jib.Contract(symbol=s, secType="STK", exchange="SMART", currency="USD")
  Jib.reqMktData(ib, idx, contract, "", false)
end

sleep(10)

for idx in 1:length(symbols)
  Jib.cancelMktData(ib, idx)
end

Jib.disconnect(ib)

```

and got this error 

```
 Info: connected
│   V = "165"
└   T = "20220207 13:44:45 Central Standard Time"
error: -1 2104 Market data farm connection is OK:uscrypto 
error: -1 2104 Market data farm connection is OK:usfuture.nj 
error: -1 2104 Market data farm connection is OK:usfarm.nj 
error: -1 2104 Market data farm connection is OK:usfuture 
error: -1 2104 Market data farm connection is OK:cashfarm 
error: -1 2104 Market data farm connection is OK:usopt 
error: -1 2104 Market data farm connection is OK:usfarm 
error: -1 2106 HMDS data farm connection is OK:ushmds.nj 
error: -1 2106 HMDS data farm connection is OK:ushmds 
error: -1 2158 Sec-def data farm connection is OK:secdefil 
marketDataType: 1 REALTIME
tickReqParams: 1 0.01 9c0001 3
tickString: 1 LAST_TIMESTAMP 1644263088
tickString: 1 LAST_EXCH Q
tickPrice2: AAPL LAST 172.81 1.0 (canAutoExecute = false, pastLimit = false, preOpen = false)
tickSize: 1 LAST_SIZE 1.0
tickSize: 1 VOLUME 519927.0
tickPrice2: AAPL HIGH 173.95 0.0 (canAutoExecute = false, pastLimit = false, preOpen = false)
tickPrice2: AAPL LOW 171.53 0.0 (canAutoExecute = false, pastLimit = false, preOpen = false)
tickPrice2: AAPL CLOSE 172.39 0.0 (canAutoExecute = false, pastLimit = false, preOpen = false)
tickPrice2: AAPL OPEN 172.76 0.0 (canAutoExecute = false, pastLimit = false, preOpen = false)
tickPrice2: AAPL BID 172.8 23.0 (canAutoExecute = true, pastLimit = false, preOpen = false)
tickPrice2: AAPL ASK 172.81 33.0 (canAutoExecute = true, pastLimit = false, preOpen = false)
tickSize: 1 BID_SIZE 23.0
tickSize: 1 ASK_SIZE 33.0
tickString: 1 BID_EXCH KPQXZU
tickString: 1 ASK_EXCH KPQZNU
marketDataType: 2 REALTIME
tickReqParams: 2 0.01 9c0001 3
┌ Error: exception thrown
│   e = TypeError: in typeassert, expected Float64, got a value of type Nothing
└ @ Jib.Reader ~/.julia/packages/Jib/SeCxV/src/reader.jl:74
[ Info: reader exiting

```

**Comment by lbilli** [comment, marked as answer]
So I converted this into a discussion, which might be a more appropriate venue.

Regarding the exception, it looks like something that I have occasionally encountered and stems from the strict type conversions that I have adopted in this package. In this case TWS is sending an empty string where a float is expected.
I could always widen the accepted types but I'm a bit reluctant doing so given the fact that usually these erratic behaviors are transient and disappear after minor TWS update.

In order to make sure that this is in fact the case, I would try first to update TWS or try a different version like "stable" or "beta".

If this behavior persists, it could be something more serious and it would be useful to have more infos, like:
- os
- TWS/IBGateway version
- account type: demo, paper, live
- hour of the day: market open, market closed, weekend

**Reply by ghost** [reply]
Hi there
thank you for transferring this to a discussion, it makes me feel a little better. I don't use github so I didn't know discussions existed. My TWS updates on boot and I do this once a week so it's the latest version. 

Here are the details you asked for.

1 OS linux mint 20.3 kernel 5.4.0-97-generic  
2 TWS build  10.12.2j    2/1/22
3 account type live
4 approx 12pm cst market open. 

 What can I do to help diagnose this please? remember I am a noob but ready to help out. If I can avoid using vscode I would prefer that but if it will help I am game to try.  How about some kind of exception handling to just throw away these issues unless they persist in a session? Could the errors be sent to a log?

This is very exciting for me and I look forwards to getting this working.

**Reply by lbilli** [reply]
First, I'd a give try to TWS Beta from [here](https://www.interactivebrokers.com/en/trading/tws.php#tws-software).

**Reply by ghost** [reply]
will do. I'll load it up on a local machine and change the ip address.

**Reply by ghost** [reply]
I loaded up the latest beta on my machine. Tried to run it and Hit a snag.

I can't run two TWS at the same time and the demo account sign up seems to have been taken away. I can't shut down the other TWS as that's been used to stream data via python.  Here's my suggestion, other people are running this and NOT hitting this issue so I'll monitor it and see how often it occurs. I posted a question on the TWS forum and the moderator checked the data from TWS and saw no problems in 44 million data items. Looks like the problem is my end. As I get more proficient in julia I'll be able to sort this out myself. Sorry to have bugged you. This is very exciting to me.

**Reply by lbilli** [reply]
Demo account username "demo_tws", no password
Or try your paper account: just check "Paper Trading" in the login screen and use your regular credentials.

**Comment by ghost** [comment]
Hi there 
I responded to an email but didn't see the event here in this area. I am new to github so I "hope" I did the right thing.  The highlevel view was that I can't use

1 the demo account approach gives 
 error: 4 354 Requested market data is not subscribed.Delayed market data is available.Error&BEST/STK/Top&BEST/STK/Top 

2 the paper money approach gives
error: 4 10197 No market data during competing live session 

 My suggestion is for me to carry on learning how to code in julia using you WONDERFUL framework and figure this out for myself. There's no need to change anything for something this transient.  The TWS forum have reminded me about the TWS logs which I completely forgot about so that something for me to consider as a double test. Compare my logs against the TWS and see what's going on.

My plan is to figure out how to log events like this to a file, have another process monitor the file and make determinations based on statistics. I would also like $12bn and a pony :-)  
thoughts?

**Reply by lbilli** [reply]
Demo account doesn't have realtime data, only delayed (as the error says). 
To get them just issue:
```julia
Jib.reqMarketDataType(ib, Jib.DELAYED)
```
before requesting market data.

**Comment by ghost** [comment]
Thanks, I thought I had done that but I didn't save the file! a REAL noob mistake.  I'm currently using the cli,text editor so I deserve everything I get but you don't sorry. 

I added the code and set the duration to one hour. It's using the TWS beta build 10.14.0u  as you recommended. Right now it's been going for 15 minutes and all is well.  After this it done I'll switch it to the TWS live machine and see what happens. It's running build 10.12.2j. 

thanks for sticking with me on this. I can only get better.  

END OF DAY UPDATE 2/8/22

**At the end of 3 runs I had NO exceptions so this supports your observation that this is a spurious event.**

 I ran it against a live system ( run 3 )  in exactly the same way I did when I got the exceptions and got none. Here are the runs I made

1 set Jib.reqMarketDataType(ib, Jib.DELAYED) run 1hour on local using TWS beta 10.14.0u and demo_tws a/c. NO EXCEPTIONS
2  set Jib.reqMarketDataType(ib, Jib.DELAYED) run 1hour on live account using TWS 10.12.2j and funded a/c. NO EXCEPTIONS
3  NOT set Jib.reqMarketDataType(ib, Jib.DELAYED) run 1hour on live account using TWS 10.12.2j and funded a/c. NO EXCEPTIONS 

I have the data runs but they don't really show anything interesting. I can post them but the live data size is 7.5mb uncompressed.  I think the best approach is for me to use this as a learning experience for julia.  Your excellent framework has given me the foundation and I am already learning. I have to work on the whole asynchronous callback functionality in julia. I get it in python but want to make sure I have the correct julia mental model.
